### PR TITLE
GELF (Graylog2) logging backend for dlog, clog and LogMonitor (was #3670)

### PR DESCRIPTION
**Pull request by irq0**
Used to be #3670. Now also integrates with clog (LogClient) and LogMonitor.

### CODE REVIEW

File: src/common/ceph_context.cc [hunk 1]
```diff
@@ -145,12 +151,38 @@ class LogObs : public md_config_obs_t {
...
       log->set_max_recent(conf->log_max_recent);
     }
+
+    // graylog
+    if (changed.count("log_to_graylog") || changed.count("err_to_graylog")) {
+      int l = conf->log_to_graylog ? 99 : (conf->err_to_graylog ? -1 : -2);
+      log->set_graylog_level(l, l);
+
+      if (conf->log_to_graylog || conf->err_to_graylog) {
+	log->start_graylog();
```
Comment (liewegas) [inline src/common/ceph_context.cc]: it's okay to call start twice?

Comment (liewegas) [inline src/common/ceph_context.cc]: looks like it.

**Comment by liewegas** [comment]
This looks good to me!  :)

**Comment by tchaikov** [comment]
seems our gitbuilders are not updated with the autobuild-ceph fdf71235, thus failed to run install-deps.sh before the build:

```
dpkg-checkbuilddeps: Unmet build dependencies: libboost-iostreams-dev (>= 1.42)
warning: dpkg-buildpackage: Build dependencies/conflicts unsatisfied; aborting.
warning: dpkg-buildpackage: (Use -d flag to override.)
```

**Comment by tchaikov** [comment]
might need to link ceph_test_cls_refcount against "libboost_iostreams", not sure why this error does not fail other builders.
```
CXXLD ceph_test_cls_refcount
./.libs/librados.so: undefined reference to `boost::iostreams::zlib::deflated'
./.libs/librados.so: undefined reference to `boost::iostreams::detail::zlib_base::before(char const*&, char const*, char*&, char*)'
./.libs/librados.so: undefined reference to `boost::iostreams::zlib_error::check(int)'
./.libs/librados.so: undefined reference to `boost::iostreams::detail::zlib_base::xdeflate(int)'
./.libs/librados.so: undefined reference to `boost::iostreams::detail::zlib_base::reset(bool, bool)'
./.libs/librados.so: undefined reference to `boost::iostreams::zlib::stream_end'
./.libs/librados.so: undefined reference to `boost::iostreams::zlib::no_flush'
./.libs/librados.so: undefined reference to `boost::iostreams::detail::zlib_base::do_init(boost::iostreams::zlib_params const&, bool, void* (*)(void*, unsigned int, unsigned int), void (*)(void*, void*), void*)'
./.libs/librados.so: undefined reference to `boost::iostreams::detail::zlib_base::zlib_base()'
./.libs/librados.so: undefined reference to `boost::iostreams::detail::zlib_base::~zlib_base()'
./.libs/librados.so: undefined reference to `boost::iostreams::zlib::default_strategy'
./.libs/librados.so: undefined reference to `boost::iostreams::detail::zlib_base::after(char const*&, char*&, bool)'
./.libs/librados.so: undefined reference to `boost::iostreams::zlib::finish'
./.libs/librados.so: undefined reference to `boost::iostreams::zlib::default_compression'
```

**Comment by dachary** [comment]
@irq0 feel free to reopen when you have time to address the comments

**Pull request closed without merging**